File: JsonToClass/PropertyModelData.m
//
//  PropertyModelData.m
//  JsonToClass
//
//  Created by pkh on 2017. 5. 31..
//  Copyright © 2017년 pkh. All rights reserved.
//

#import "PropertyModelData.h"
#import "NSString+Helper.h"



@implementation PropertyModelData

+ (PropertyModelData*)allocWithKey:(NSString*)key object:(NSObject*)object perfix:(NSString*)perfix {
    return [[PropertyModelData alloc] initWithKey:key object:object perfix:perfix];
}

- (id)initWithKey:(NSString*)key object:(NSObject*)object perfix:(NSString*)perfix {
    self = [super init];
    if(self)
    {    
        [self setKey:key object:object perfix:perfix];
    }
    return self;
}

- (void)setKey:(NSString*)key object:(NSObject*)object perfix:(NSString*)perfix {
    self.key = key;
    self.value = object;
    self.perfix = perfix;
}

- (NSString*)getStringObjectC {
    
    NSString *directive = @"strong";
    NSString *className = nil;
    NSString *keyName = [NSString stringWithFormat:@"*%@", self.key];
    
    if ([self.value isKindOfClass:[NSString class]])
    {
        className = @"NSString";
    }
    else if ([self.value isKindOfClass:NSNumber.class])
    {
        directive = @"assign";
        keyName = self.key;
        className = [self getNumberClassNumberType:(NSNumber*)self.value];
    }
    else if ([self.value isKindOfClass:NSArray.class])
    {
        NSArray *obj = (NSArray*)self.value;
        NSObject *subObj = obj.firstObject;
        if ([subObj isKindOfClass:NSString.class]) {
            className = @"NSArry<NSString*>";
        }
        else {
            className = [NSString stringWithFormat:@"NSMutableArray<%@%@%@*>",self.perfix, [self.key upercaseFirstChar], ARRAY_INNER_CLASS_TAIL_PIX];
        }
    }
    else if ([self.value isKindOfClass:[NSDictionary class]])
    {
        className = [NSString stringWithFormat:@"%@%@", self.perfix, [self.key upercaseFirstChar]];
    }
    else if ([self.value isKindOfClass:[NSNull class]])
    {
        className = @"unknow";
    }
    
    return [NSString stringWithFormat:@"\n@property (nonatomic, %@) %@ %@;",directive, className, keyName];
    
}

- (NSString*)getStringSwif {
    
    NSString *className = nil;
    NSString *defaultValue = @" = 0";
    
    if ([self.value isKindOfClass:NSString.class])
    {
        className = @"String";
        defaultValue = @" = \"\"";
        return [NSString stringWithFormat:@"\n    var %@: %@%@", self.key, className, defaultValue];
    }
    else if ([self.value isKindOfClass:NSNumber.class])
    {
        className = [self getNumberClassNumberTypeSwift:(NSNumber*)self.value];
        defaultValue = [self getNumberDefaultValueNumberTypeSwift:(NSNumber*)self.value];
        return [NSString stringWithFormat:@"\n    var %@: %@%@", self.key, className, defaultValue];
    }
    else if ([self.value isKindOfClass:NSArray.class])
    {
        NSArray *obj = (NSArray*)self.value;
        NSObject *subObj = obj.firstObject;
        if ([subObj isKindOfClass:NSString.class]) {
            return [NSString stringWithFormat:@"\n    var %@ = [String]()", self.key];
        }
        else {
            className = [NSString stringWithFormat:@"[%@%@%@]",self.perfix, [self.key upercaseFirstChar], ARRAY_INNER_CLASS_TAIL_PIX];
            defaultValue =[NSString stringWithFormat:@"%@()", className];
            return [NSString stringWithFormat:@"\n    var %@ = %@", self.key, defaultValue];
        }
        
    }
    else if ([self.value isKindOfClass:NSDictionary.class])
    {
        className = [NSString stringWithFormat:@"%@%@", self.perfix, [self.key upercaseFirstChar]];
        defaultValue = [NSString stringWithFormat:@"%@?", className];
        return [NSString stringWithFormat:@"\n    var %@: %@", self.key, defaultValue];
    }
    else if ([self.value isKindOfClass:NSNull.class])
    {
        className = @"unknow??";
        return [NSString stringWithFormat:@"\n    var %@: %@", self.key, className];
    }
    else {
        return @"";
    }
    
    
    
}


- (NSString*)getNumberClassNumberType:(NSNumber*)number
{
    //            NSLog(@"key = %@, value = %@, value class = %ld", key, value, numberType);
    switch (CFNumberGetType((CFNumberRef)number))
    {
        case kCFNumberSInt64Type:
            return @"NSInteger";
            break;
        case kCFNumberCharType:
            return @"BOOL";
            break;
        case kCFNumberFloat64Type:
            return @"CGFloat";
            break;
        default:
            return @"NSNumber";
            break;
    }
}

- (NSString*)getNumberClassNumberTypeSwift:(NSNumber*)number
{
    //            NSLog(@"key = %@, value = %@, value class = %ld", key, value, numberType);
    switch (CFNumberGetType((CFNumberRef)number))
    {
        case kCFNumberSInt64Type:
            return @"Int";
            break;
        case kCFNumberCharType:
            return @"Bool";
            break;
        case kCFNumberFloat64Type:
            return @"Float";
            break;
        default:
            return @"NSNumber";
            break;
    }
}


- (NSString*)getNumberDefaultValueNumberTypeSwift:(NSNumber*)number
{
    //            NSLog(@"key = %@, value = %@, value class = %ld", key, value, numberType);
    switch (CFNumberGetType((CFNumberRef)number))
    {
        case kCFNumberSInt64Type:
            return @" = 0";
            break;
        case kCFNumberCharType:
            return @" = false";
            break;
        case kCFNumberFloat64Type:
            return @" = 0.0";
            break;
        default:
            return @" = 0";
            break;
    }
}

- (NSString*)isImportClass {
    NSString *className = nil;
    
    if ([self.value isKindOfClass:NSArray.class])
    {
        className = [NSString stringWithFormat:@"\n#import \"%@%@%@.h\"",self.perfix, [self.key upercaseFirstChar], ARRAY_INNER_CLASS_TAIL_PIX];
    }
    else if ([self.value isKindOfClass:[NSDictionary class]])
    {
        className = [NSString stringWithFormat:@"\n#import \"%@%@.h\"", self.perfix, [self.key upercaseFirstChar]];
    }
    
    return className;
}

@end

File: swift/JsonToClass/PropertyModelData.swift
//
//  PropertyModelData.swift
//  JsonToClass
//
//  Created by pkh on 17/04/2019.
//  Copyright © 2019 pkh. All rights reserved.
//

import Foundation

let ARRAY_INNER_CLASS_TAIL_PIX = "Item"


func numberType(_ value: Any?) -> Any? {
    if let number = value as? NSNumber {
        switch CFNumberGetType(number) {
        case .charType: //Bool
            return value as! Bool
        case .sInt8Type, .sInt16Type, .sInt32Type, .sInt64Type, .shortType, .intType, .longType, .longLongType, .cfIndexType, .nsIntegerType: //Int
            return value as! Int
        case .float32Type, .float64Type, .floatType, .doubleType, .cgFloatType: //Double
            return value as! CGFloat
        default:
            return value
        }
    }
    return value
}

class PropertyModelData {
    var value: Any?
    var key = ""
    var perfix = ""
    
    init(key: String, object: Any?, perfix: String) {
        setKey(key, object: object, perfix: perfix)
    }
    
    func setKey(_ key: String, object: Any?, perfix: String) {
        self.key = key
        self.value = object
        self.perfix = perfix
    }
    
    func getStringObjectC() -> String {
        
        var directive = "strong"
        var className: String = ""
        var keyName = "*\(key)"
        
        if numberType(value) is Int {
            directive = "assign"
            keyName = key
            className = "NSInteger"
        }
        else if numberType(value) is CGFloat {
            directive = "assign"
            keyName = key
            className = "CGFloat"
        }
        else if numberType(value) is Bool {
            directive = "assign"
            keyName = key
            className = "BOOL"
        }
        else if value is String {
            className = "NSString"
        }
        else if let obj = value as? [Any] {
            if (obj.first is String) {
                className = "NSMutableArray<NSString*>"
            } else {
                className = "NSMutableArray<\(perfix)\(key.capitalizedFirst())\(ARRAY_INNER_CLASS_TAIL_PIX)*>"
            }
        }
        else if value is [String : Any] {
            className = "\(perfix)\(key.capitalizedFirst())"
        }
        else  {
            className = "unknow"
        }
        
        return "\n@property (nonatomic, \(directive)) \(className) \(keyName);"
        
    }
    
    func getStringSwif() -> String {
        
        if numberType(value) is Int {
            return "\n    var \(key): Int = 0"
        }
        else if numberType(value) is CGFloat {
            return "\n    var \(key): CGFloat = 0.0"
        }
        else if numberType(value) is Bool {
            return "\n    var \(key): Bool = false"
        }
        else if numberType(value) is String {
            return "\n    var \(key): String = \"\""
        }
        else if let obj = value as? [Any] {
            if (obj.first is String) {
                return "\n    var \(key) = [String]()"
            } else {
                return "\n    var \(key) = [\(perfix)\(key.capitalizedFirst())\(ARRAY_INNER_CLASS_TAIL_PIX)]()"
            }
        }
        else if value is [String : Any] {
            return "\n    var \(key): \(perfix)\(key.capitalizedFirst())?"
        }
        else {
            return "\n    var \(key): unknow??"
        }
    }
    

    func isImportClass() -> String? {
        var className: String? = nil
        
        if let obj = value as? [Any], obj.count > 0 {
            let item = obj.first as? NSObject
            if item is [String : Any] {
                className = "\n#import \"\(perfix)\(key.capitalizedFirst())\(ARRAY_INNER_CLASS_TAIL_PIX).h\""
            }
        } else if (value is [String : Any]) {
            className = "\n#import \"\(perfix)\(key.capitalizedFirst()).h\""
        }
        
        return className
    }

}
